src/ra_server.erl

```diff
@@ -3314,6 +3443,186 @@ after_log_append_reply(Cmd, Idx, Term, Effects0) ->
+                       fun
+                           (PeerId, _, Acc) when PeerId =:= Id ->
+                               Acc;
+                           (PeerId, _, Acc) ->
+                               Acc ++ info_rpc_effects_for_peer(State, PeerId)
+                       end, [], Cluster),
+    InfoRpcEffects.
+
+info_rpc_effects_for_peer(
+  #{cluster := Cluster, current_term := CurTerm} = State, PeerId) ->
+    %% We determine if we need to ask (for the fist time or again) the info
+    %% from a peer.
+    SendRpc = case Cluster of
+                  #{PeerId := #{machine_version := PeerMacVer}} ->
+                      %% We have the machine version of the peer, but we want
+                      %% to ask again if that version is old, in the hope the
+                      %% peer restarted and was updated.
+                      MacVer = machine_version(State),
+                      PeerMacVer < MacVer;
+                  _ ->
+                      %% We don't have any details about the peer, we ask.
+                      true
+              end,
+    case SendRpc of
+        true ->
+            %% We ask for all info keys currently. If we ask for specific
+            %% keys, we will have to handle merging of already known info keys
+            %% and updates.
+            Id = id(State),
+            Command = #info_rpc{from = Id,
+                                term = CurTerm,
+                                keys = [machine_version]},
+            [{send_rpc, PeerId, Command}];
+        false ->
+            []
+    end.
+
+info_reply_effect(
+  #{current_term := CurTerm} = State,
+  #info_rpc{from = FromId, keys = Keys}) ->
+    Id = id(State),
+    Info = ra_server_info(State, Keys),
+    InfoReply = #info_reply{from = Id,
+                            term = CurTerm,
+                            keys = Keys,
+                            info = Info},
+    {cast, FromId, InfoReply}.
+
+empty_info_reply_effect(
+  #{current_term := CurTerm} = State,
+  #info_rpc{from = FromId, keys = Keys}) ->
+    Id = id(State),
+    InfoReply = #info_reply{from = Id,
+                            term = CurTerm,
+                            keys = Keys,
+                            info = undefined},
+    {cast, FromId, InfoReply}.
+
+ra_server_info(State) ->
+    MacVer = machine_version(State),
+    #{machine_version => MacVer}.
+
+ra_server_info(State, Keys) ->
+    %% Note that the node that asked may ask for keys we don't know.
+    Info0 = ra_server_info(State),
+    Info1 = maps:filter(
+              fun(Key, _Value) ->
+                      lists:member(Key, Keys)
+              end, Info0),
+    Info1.
+
+handle_info_reply(
+  #{cluster := Cluster} = State,
+  #info_reply{from = PeerId, keys = Keys, info = Info}) ->
+    PeerState0 = maps:get(PeerId, Cluster),
+    PeerState1 = case Info of
+                     undefined ->
+                         %% The peer is in a Raft state where it can't give
```

Comment:
I think a peer should always be able to provide info for keys it understands.

Reply:
Ok, so I should get rid of the `empty_info_reply_effect()` function and let a Ra server always send its machine version in the `#info_reply{}`, correct?

Reply:
yes I think that makes sense,

Reply:
In fact, disregard that previous comment.

Reply:
Addressed too :)